doc/design/backfill-starvation-prevention.md

```diff
@@ -0,0 +1,179 @@
+
+![backfill-use-case](../images/backfill-use-case.png)
+
+### Starvation Prevention
+The following diagram illustrates why we need to add starvation prevention capability into kube batch scheduler.
+
+Suppose task A requires 2 CPUs. In the first round, scheduler notices that the node does not have enough CPU to accommodate task A, therefore the scheduler will skip task A, and allocate CPU1 to task B. In the second round, suppose task B occupies CPU1, and CPU2 becomes available.  In this case, we still only have one idle CPU, therefore task A will be skipped again. Scheduler will allocate the node to task C. If this pattern goes on, task A will be starved. Preventing task A from being starved is another issue we need to address in this project.
+
+![starvation-use-case](../images/starvation-use-case.png)
+
+## Objectives
+* Improve Resource Utilization with Backfill
+  * When we schedule jobs by their priorities, we may run into situations where resources are allocated to tasks of a job which is not ready to run. This behavior leads to suboptimal resource utilization.
+  * One goal of this project is to improve resource utilization with backfill - to run tasks in those allocated, but unused, resources.
+  * If preemption is enabled, backfill tasks will be preempted whenever the job (which lends resource) is ready to run.
+  * If preemption is disabled, the job (which lends resource) will wait for backfill tasks to finish and then start.
+
+* No Starvation
+  * A job’s wait time should be bounded if 1) the amount of resources requested by the job is less the cluster’s total capacity, and 2) no continuous arrival of higher priority jobs.
+  * It is acceptable for a job to wait forever if the amount of its requested resources is greater than the total capacity of the cluster. This issue will be addressed by cluster auto scaling.
+  * It is acceptable for a job to wait forever if jobs with higher priority keep coming. In this case, the customer needs to increase the priority of the starving job.
+
+## Terminologies
+This design document introduces the following new terminologies.
+
+* Starving Job
+  
+  A job that has been in pending state longer than starvation threshold. Starvation threshold is a scheduler configuration.
+
+* Backfill Task
+  
+  Tasks run inside resources that are allocated to other tasks.
+
+* Backfilled Resource
+
+  Resources on a node that are occupied by backfill tasks
+  
+* Accessible Resource
+
+  Accessible Resources of a node is the sum of the node’s idle resource and the node’s backfilled resource
+
+* Allocated Task
+
+  Task allocated with idle resources
+  
+* Conditionally Allocated Task
+
+  Task allocated with backfilled resources
```

Comment:
Need we take preemptor task as conditionally allocated task?

Reply:
A conditionally allocated task is a task that has been allocated with *accessible* resources which include idle resources and backfilled resources. Idle resources are ready to use. Backfilled resources are resources that are currently occupied by some backfill tasks. After backfill tasks exit (either voluntarily or being preempted), the conditionally allocated task should have enough resource to start. In order words, the task will have enough resources if backfill tasks exit. 

Preemptors are *conditionally ready* jobs. A job is considered as conditionally ready if it has enough conditionally allocated tasks. Conditionally ready jobs will have enough resources to start if backfill tasks exit.